[WIP] Node compatibility

LeartS opened the pull request:
This is a pull request to make strophe.js able to run in nodejs.

This pull request is still work in progress and not ready for merge, I've opened it so that we can start a discussion on this and others may be able to pitch in.

## Why

Strophejs is a very well written library, has an API nicer than most other xmpp node libraries, has [excellent external type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/strophe.js), has a nice plugin system, and the browser-specific stuff is already limited; therefore I think it's would be great to add node compatibility to be able to use strophe.js on the backend, for example to write xmpp bots.

## What's done

* Fix [webpack bug](https://github.com/webpack/webpack/issues/6525) that produced broken universal builds that required `window`
* Use `isomorphic-ws` to provide an universal websocket module that uses the native WebSocket object on browser and the [ws](https://www.npmjs.com/package/ws) package on node.
At the moment I have a problem with webpack only bundling the `node.js` or `browser.js` entrypoints of `isomorphic-ws`, and not both, so I think websockets under browsers are broken until I fix that 
* Use `xmldom` to provide `DOMParser`, `document` and the other DOM API used by `Strophe.Builder` to build xml stanzas.
* Use `abab` to provide a cross-platform standard-adhering implementation of `atob` and `btoa` used by `Strophe.Websocket`

Regarding the additional dependencies I've introduced:

* `abab` and `isomorphic-ws` are very small packages without dependencies, so I've added them directly.
* `ws` and `xmldom` are larger packages, but they are defined as peerDependency and optionalDependency, so they won't be installed automatically but it's the responisibility of users to installthose packages when using strophe.js under node.

This way the strophe.js bundle size remains pretty much the same. 

At this point, I already got strophe working using websockets in node.

## What's missing

### BOSH

At the moment I've not worked on BOSH yet. I was thinking as a first version, strophe.js could be compatibile with node only when using the websocket protocol, and later also add BOSH compatibility.

What's needed for BOSH:
* XmlHTTPRequest
* cookies

### Tests

jcbrand commented:
This is a much requested feature, so thanks very much for working on it.

Why did you remove the `dist` files?

> ws and xmldom are larger packages, but they are defined as peerDependency and optionalDependency

I don't see an actual peerDependency being defined in package.json.

LeartS commented:
Regarding dependencies:

* `ws` is a `peerDependency` of `isomorphic-ws`, which is a dependency of this package.
* `xmldom` is a optional dependency of this package.

Regarding dist:

dist contains builded files, usually they should not be tracked in git.
I was planning on putting a [`prepare` script](https://docs.npmjs.com/misc/scripts#prepublish-and-prepare) on the package.json which runs `build`, and then publish only the contents of `dist/` + package metadata on NPM, and only the source files on git.

jcbrand commented:
> I was planning on putting a prepare script on the package.json which runs build, and then publish only the contents of dist/ + package metadata on NPM, and only the source files on git.

I was hoping you were going to say this. I heard somewhere that this was possible but didn't yet find out how to do it.

LeartS commented:
You can do it by combining these two NPM features:
* You can use the [`files` field of package.json](https://docs.npmjs.com/files/package.json#files) to tell NPM which files to include in the packaged version.
* If you put a `prepare` script in the `scripts` key of `package.json`, that script is run automatically before publishing on NPM and also when installing as a dependency from a git url; which means that for installations from npm registry ( `npm add --save strophe.js` ) or from github (`npm add --save github:strophe/strophe.js#branch`) users will have the built version automatically.

I'll create a new PR for this packaging changes so that it can be processed indipendently of the nodejs compatibility changes.

LeartS commented:
Rebased and removed dependency on `isomorphic-ws`. Turns out it didn't basically do anything, as `ws` should be comptabile with the WebSocket standard by itself, so I just used a conditional require like for DOMParser.

jcbrand commented:
@LeartS are you done with this PR?

LeartS commented:
Hi @jcbrand 

I'm using the branch to build an higher-level xmpp node library and it seems to work well using websockets. I can connect, disconnect, send presences, receive/send messages etc. I also uses some strophe plugins (ping, roster, mam) and it seems like they all work correctly with the modified strophe.

I haven't yet tested it with BOSH on node, nor have I tested this modified version in the browser to check that I didn't break anything, I was having some problems with the testing infrastructure.

If this PR doesn't break the browser version (it shouldn't) I think it can be merged as an initial nodejs support. It's probably not perfect and completely untested with BOSH, but better than no support at all; and it can be improved. Or maybe you want to keep this node version in a parallel branch for a bit, people who want to use it could install it using github:strophe/strophejs#node

sualko commented on src/bosh.js:
```diff
@@ -11,6 +11,16 @@ import core from 'core';
 
 const Strophe = core.Strophe;
 const $build = core.$build;
+let nodeXHR;
```
Why can this not be moved to the `_newXHR` function?

LeartS replied:
I honestly don't remember, I think I tried putting it there but the linter or even babel was complaining about having a require in the method.

sualko commented on src/core.js:
```diff
@@ -6,10 +6,20 @@
 import SHA1 from 'sha1';
 import utils from 'utils';
 
+let xmldom;
```
Same here. Why can't it moved to `_makeGenerator`? I think the readability is degraded. Does it maybe make sense to put all of this feature detection stuff to it's own polyfill file?

LeartS replied:
Same as `nodeXHR`. :+1: for putting feature detection and conditional imports in a separate module (file), I just have to check because I had a lot of issues with conditional imports + babel + webpack.

sualko commented on tests/index.html:
```diff
@@ -1,24 +1,23 @@
 <html>
   <head>
-	<script src="../node_modules/qunitjs/qunit/qunit.js"></script>
+    <script src="../node_modules/qunitjs/qunit/qunit.js"></script>
```
Is there any reason the format was changed?

LeartS replied:
Just for consistency. It was a mix of tabs and spaces, I converted all to spaces.

jcbrand commented:
You might have noticed that I recently made a new release of Strophe.js.

I think it would be a good idea to merge this branch soon, that'll give us some time to test it out without having any pressure to make another release.

LeartS commented:
Sure! What do you want me to do for the release?

jcbrand commented:
> Sure! Do you need me to do something in preparation for the release?

Nothing comes to mind right now.

LeartS commented:
I'm working on a rewrite of this pull request that depends on and works with https://github.com/strophe/strophejs/pull/328 . It should be ready by today.

LeartS commented:
I'm closing this in favour of #329
